File: es7/mynlp-plugin/src/main/java/com/mayabot/mynlp/es/MynlpPlugin.java
/*
 *  Copyright 2017 mayabot.com authors. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package com.mayabot.mynlp.es;

import com.google.common.collect.Lists;
import com.mayabot.mynlp.es.nlpdata.NlpSchedulerJog;
import com.mayabot.nlp.MynlpBuilder;
import com.mayabot.nlp.Mynlps;
import org.apache.lucene.analysis.Analyzer;
import org.elasticsearch.SpecialPermission;
import org.elasticsearch.client.Client;
import org.elasticsearch.cluster.service.ClusterService;
import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
import org.elasticsearch.common.settings.Setting;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.common.xcontent.NamedXContentRegistry;
import org.elasticsearch.env.Environment;
import org.elasticsearch.env.NodeEnvironment;
import org.elasticsearch.index.analysis.AnalyzerProvider;
import org.elasticsearch.index.analysis.TokenizerFactory;
import org.elasticsearch.indices.analysis.AnalysisModule;
import org.elasticsearch.plugins.AnalysisPlugin;
import org.elasticsearch.plugins.Plugin;
import org.elasticsearch.plugins.SearchPlugin;
import org.elasticsearch.script.ScriptService;
import org.elasticsearch.search.fetch.FetchSubPhase;
import org.elasticsearch.threadpool.ThreadPool;
import org.elasticsearch.watcher.ResourceWatcherService;

import java.nio.file.Path;
import java.util.*;
import java.util.function.Consumer;

/**
 * @author jimichan
 */
public class MynlpPlugin extends Plugin implements AnalysisPlugin, SearchPlugin {

    /**
     * cws模型比较重，默认不开启。
     */
    private boolean enableCws;
    private String mynlpServer;


    public MynlpPlugin(Settings settings, Path configPath) {

        SecurityManager sm = System.getSecurityManager();
        if (sm != null) {
            sm.checkPermission(new SpecialPermission());
        }

        enableCws =  MynlpPluginSettings.getEnableCws().get(settings);
        mynlpServer = MynlpPluginSettings.getServer().get(settings);

        Mynlps.install(builder -> {
            builder.set("mynlp.server",mynlpServer);
        });
    }

    @Override
    public Collection<Object> createComponents(Client client, ClusterService clusterService, ThreadPool threadPool, ResourceWatcherService resourceWatcherService, ScriptService scriptService, NamedXContentRegistry xContentRegistry, Environment environment, NodeEnvironment nodeEnvironment, NamedWriteableRegistry namedWriteableRegistry) {
        //
        // 这里可以监控NlpService。每隔1分钟轮训一次。查看是否更新了资源。
        // client.n
        // nodeEnvironment.nodeId()
        // 然后把同步过程的日志信息写入固定的索引，其他系统可以在日志里面查询到每个节点的数据同步情况。
        //

        NlpSchedulerJog job = new NlpSchedulerJog(threadPool, nodeEnvironment.nodeId());

        return Lists.newArrayList(job);
    }

    @Override
    public List<Setting<?>> getSettings() {
        return Arrays.asList(
                MynlpPluginSettings.getServer(),
                MynlpPluginSettings.getEnableCws());
    }


    @Override
    public Map<String, AnalysisModule.AnalysisProvider<TokenizerFactory>> getTokenizers() {
        Map<String, AnalysisModule.AnalysisProvider<TokenizerFactory>> extra = new HashMap<>();

        extra.put("mynlp", MynlpTokenizerFactory::new);
        extra.put("mynlp-core", MynlpTokenizerFactory::new);

        if (enableCws) {
            extra.put("mynlp-cws", MynlpTokenizerFactory::new);
        }

        extra.put("issue", ChineseIssueTokenizerFactory::new);
        extra.put("hash", StringHashTokenizerFactory::new);

        extra.put("pinyin", PinyinTokenizerFactory::new);
        extra.put("pinyin-head", PinyinTokenizerFactory::new);
        extra.put("pinyin-stream", PinyinTokenizerFactory::new);
        extra.put("pinyin-fuzzy", PinyinTokenizerFactory::new);
        extra.put("pinyin-keyword", PinyinTokenizerFactory::new);
        extra.put("pinyin-head-keyword", PinyinTokenizerFactory::new);
        extra.put("pinyin-fuzzy-keyword", PinyinTokenizerFactory::new);

        return extra;
    }


    @Override
    public Map<String, AnalysisModule.AnalysisProvider<AnalyzerProvider<? extends Analyzer>>> getAnalyzers() {

        //开启异步线程，先执行一个core的分词，试图去下载依赖的资源

        Map<String, AnalysisModule.AnalysisProvider<AnalyzerProvider<? extends Analyzer>>> map = new HashMap<>(10);

        map.put("issue", ChineseIssueNumProvider::new);
        map.put("hash", StringHashProvider::new);

        map.put("pinyin", PinyinProvider::new);
        map.put("pinyin-head", PinyinProvider::new);
        map.put("pinyin-stream", PinyinProvider::new);
        map.put("pinyin-fuzzy", PinyinProvider::new);
        map.put("pinyin-keyword", PinyinProvider::new);
        map.put("pinyin-head-keyword", PinyinProvider::new);
        map.put("pinyin-fuzzy-keyword", PinyinProvider::new);


        map.put("mynlp", MynlpAnalyzerProvider::new);
        map.put("mynlp-core", MynlpAnalyzerProvider::new);
        if (enableCws) {
            map.put("mynlp-cws", MynlpAnalyzerProvider::new);
        }

        return map;
    }

    @Override
    public List<SearchExtSpec<?>> getSearchExts() {
        return Lists.newArrayList(SummaryExtKt.regSummarySearchExtBuilder());
    }

    @Override
    public List<FetchSubPhase> getFetchSubPhases(FetchPhaseConstructionContext context){
        return Lists.newArrayList(new SummaryFetchSubPhase());
    }
}

File: plugin/plugin-main/src/main/java/com/mayabot/mynlp/MynlpPlugin.kt
package com.mayabot.mynlp

import com.mayabot.nlp.Mynlp
import org.apache.lucene.analysis.Analyzer
import org.elasticsearch.SpecialPermission
import org.elasticsearch.common.logging.Loggers
import org.elasticsearch.common.settings.Setting
import org.elasticsearch.common.settings.Settings
import org.elasticsearch.index.analysis.AnalyzerProvider
import org.elasticsearch.index.analysis.TokenizerFactory
import org.elasticsearch.indices.analysis.AnalysisModule
import org.elasticsearch.plugins.AnalysisPlugin
import org.elasticsearch.plugins.Plugin
import org.elasticsearch.plugins.SearchPlugin
import java.nio.file.Path
import java.nio.file.Paths
import java.security.AccessController
import java.security.PrivilegedAction

class MynlpPlugin(
    val settings: Settings, val configPath: Path
) : Plugin(), SearchPlugin, AnalysisPlugin{

    private val mynlpServer:String

    private val mynlp: Mynlp

    private val dmsService:DmsService

    init {
        logger.info("start load MynlpManaPlugin ...")
        System.getSecurityManager()?.let {
            AccessController.checkPermission(SpecialPermission())
        }

        Mynlp.configer().setDataDir(
            Paths.get(configPath.toString(),"mynlp.data").toString()
        )

        Mynlp.configer().setCacheDir(
            Paths.get(System.getProperty("java.io.tmpdir"),"mynlp.temp").toString()
        )

        mynlpServer = ExtSettings.server.get(settings)

        this.dmsService = AccessController
            .doPrivileged(
                PrivilegedAction<DmsService>{
                    val mynlp = Mynlp.instance()
                    DmsService(HttpMynlpResourceService(mynlpServer),mynlp)
                })

        this.mynlp = dmsService.mynlp

        logger.info("Mynlp Plugin init ok")
    }

    override fun getTokenizers(): MutableMap<String, AnalysisModule.AnalysisProvider<TokenizerFactory>> {
        val extra =  HashMap<String, AnalysisModule.AnalysisProvider<TokenizerFactory>>()

        extra["mynlp"] = MynlpDmsAnalysisProvider(dmsService,::MynlpTokenizerFactory)
        extra["mynlp_index_atom"] = MynlpDmsAnalysisProvider(dmsService,::MynlpTokenizerFactory)
        extra["mynlp_index_overlap"] = MynlpDmsAnalysisProvider(dmsService,::MynlpTokenizerFactory)
        extra["mynlp_smart_atom"] = MynlpDmsAnalysisProvider(dmsService,::MynlpTokenizerFactory)
        extra["mynlp_smart_overlap"] = MynlpDmsAnalysisProvider(dmsService,::MynlpTokenizerFactory)

        return extra
    }

    override fun getAnalyzers(): MutableMap<String, AnalysisModule.AnalysisProvider<AnalyzerProvider<out Analyzer>>> {
        val extra =  HashMap<String, AnalysisModule.AnalysisProvider<AnalyzerProvider<out Analyzer>>>()

        extra["mynlp"] = MynlpDmsAnalysisProvider(dmsService,::MynlpAnalyzerProvider)
        extra["mynlp_index_atom"] = MynlpDmsAnalysisProvider(dmsService,::MynlpAnalyzerProvider)
        extra["mynlp_index_overlap"] = MynlpDmsAnalysisProvider(dmsService,::MynlpAnalyzerProvider)
        extra["mynlp_smart_atom"] = MynlpDmsAnalysisProvider(dmsService,::MynlpAnalyzerProvider)
        extra["mynlp_smart_overlap"] = MynlpDmsAnalysisProvider(dmsService,::MynlpAnalyzerProvider)

        return extra
    }

    /**
     * 返回插件自定义的配置项
     * 如果不明确指明settings，那么es启动时会校验报错
     */
    override fun getSettings(): MutableList<Setting<*>> {
        return ArrayList<Setting<*>>().apply {
            ExtSettings.server
        }
    }

    companion object{
        val logger = Loggers.getLogger(MynlpPlugin::class.java,"mynlp")
    }

    object ExtSettings {
        val server: Setting<String> = Setting.simpleString("mynlp.server", "", Setting.Property.NodeScope)
    }
}